### PR TITLE
perf(neighbors): tile partial naive neighbor lists

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Accelerate native, JAX, and Torch naive neighbor-list rebuilds that update a compact subset of atoms. The previous implementation used scalar traversal in framework partial paths even when a CUDA tile-cooperative topology-only specialization was available, making large-system partial rebuilds disproportionately expensive. Expose the native specialization consistently through JAX and Torch, preserve scalar behavior for unsupported outputs, and retain optional JAX-only Warp graph replay for callers that keep all graph inputs and outputs at stable addresses.

The new compact-row path maps each output row to its global source atom:

```python
ishift, row = wp.tid()
atom_i = row
if PARTIAL:
    atom_i = target_indices[row]
```

For batched selective rebuilds, reset compact output rows using the target atom's owning system rather than the compact-row index:

```python
row = wp.tid()
atom = target_indices[row]
if rebuild_flags[batch_idx[atom]]:
    num_neighbors[row] = 0
```

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

N/A

## Changes Made

- Add CUDA tiled naive neighbor-matrix support for topology-only compact `target_indices` rows, including no-PBC, wrapped PBC, and prewrapped PBC paths.
- Expose the same topology-only partial tile path through JAX and Torch single-system APIs; expose explicit tile through both batched APIs while retaining scalar batched `strategy="auto"`.
- Add dtype-specific single-system automatic dispatch thresholds: tile at 256 atoms for float64 and 1,024 atoms for float16/float32; retain scalar automatic dispatch for batched partial and batched prewrapped PBC workloads.
- Fix batched PBC strategy propagation, scalar partial-PBC wrapping parity, target-aware selective reset behavior, and tiled-kernel input ABI coverage.
- Add JAX opt-in `graph_mode="warp"` replay for tiled partial topology calls with persistent caller-owned buffers; reject dynamic COO output in this mode.
- Add regression coverage for scalar/tile count equality and order-independent `(neighbor, periodic-shift)` parity, graph configuration isolation, persistent-graph constraints, batched routing, and target-aware resets.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

Tile execution is CUDA-only and intentionally limited to topology-only results. Requests for distances, vectors, pair functions, or selective rebuilds continue to use scalar kernels. Scalar and tiled neighbor rows may have different entry order; their active neighbor and periodic-shift multisets are equivalent. The benchmark numbers below are JAX measurements; this PR makes no Torch performance claim.

Controlled RTX 6000 Ada benchmarks used a 9,660-atom periodic system, six target rows, a 12 Å cutoff, and a capacity of 4,096 neighbors per row. Isolated synchronized neighbor-list calls measured:

| Dtype | Scalar | Tile | Speedup |
| --- | ---: | ---: | ---: |
| float32 | 1,266.2 µs | 234.8 µs | 5.39x |
| float64 | 4,887.7 µs | 257.8 µs | 18.96x |

Every scalar/tile run returned 2,074 active edges with exact count equality and exact order-independent `(neighbor, periodic-shift)` multiset equality.

With stable pointers, optional Warp graph replay reduced isolated tiled partial-call latency from 228.4 to 125.2 µs for float32 (1.82x) and from 255.2 to 153.5 µs for float64 (1.66x). Fresh input addresses made graph replay slower, so graph replay remains opt-in and requires persistent buffers.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/neighbors/naive/kernels.py [hunk 1]
```diff
@@ -999,14 +1003,18 @@ def _kernel(
...
 
         See Also
         --------
         get_naive_neighbor_matrix_kernel : Return the specialized naive tile-output kernel.
         """
-        ishift, atom_i = wp.tid()
+        ishift, row = wp.tid()
+        atom_i = row
+        if PARTIAL:
+            atom_i = target_indices[row]
```
Comment (laserkelvin) [inline nvalchemiops/neighbors/naive/kernels.py:1017]: blocking: `target_indices[row]` is used immediately for position and batch lookups, but the public wrappers only validate rank, dtype, and device; an out-of-range value can become an illegal device read. I assume invalid public input must fail safely. Add an on-device bounds guard before indexed reads (and an eager `ValueError` where feasible), with negative/too-large tests; handle uniqueness separately.

File: test/neighbors/bindings/jax/test_naive.py [hunk 2]
```diff
@@ -890,6 +1216,257 @@ def _make_naive_stale_inputs(
...
+                    num_neighbors=nn,
+                    strategy=strategy,
+                    graph_mode="warp",
+                )
+
+            graph_result = graph_step(positions, neighbor_matrix, num_neighbors)
+
+        jax.block_until_ready(graph_result)
+        first_result = tuple(np.asarray(array) for array in graph_result)
+        graph_result = graph_step(positions, *graph_result)
```
Comment (laserkelvin) [inline test/neighbors/bindings/jax/test_naive.py:1346]: issue: This replay uses the same `positions` as capture, so it only checks reset/idempotence; stale captured geometry or wrapping state would still pass. I assume replay must follow changing MD coordinates. Replay with positions that change a target's neighbor set—ideally across PBC—and compare each step with `graph_mode="none"`.

**Review by laserkelvin — COMMENTED** [review]
Thanks for the thorough cross-backend coverage. I left nine focused inline comments covering safety, behavior and regression gaps, auto-dispatch, API consistency, and documentation.

File: test/neighbors/bindings/jax/test_batch_naive.py [hunk 3]
```diff
@@ -367,20 +392,154 @@ def test_target_indices_rejects_full_size_user_buffers(self):
...
-    def test_target_indices_rejects_tile_strategy(self):
-        """Explicit tiled batch naive mode does not support partial rows."""
-        positions = jnp.array([[0.0, 0.0, 0.0], [0.5, 0.0, 0.0]], dtype=jnp.float32)
-        batch_idx = jnp.array([0, 0], dtype=jnp.int32)
-        with pytest.raises(NotImplementedError, match="target_indices"):
+    @pytest.mark.gpu
+    @pytest.mark.parametrize("dtype", [jnp.float32, jnp.float64])
+    @pytest.mark.parametrize("half_fill", [False, True])
+    @pytest.mark.parametrize("pbc_mode", ["none", "wrapped", "prewrapped"])
+    def test_target_indices_tile_matches_scalar(self, dtype, half_fill, pbc_mode):
```
Comment (laserkelvin) [inline test/neighbors/bindings/jax/test_batch_naive.py:403]: issue: The new batched partial coverage exercises topology only, leaving the compact JAX geometry/pair-output path untested. I assume reversed targets must select the correct per-system cell and preserve pair gradients. Add a non-empty PBC case with distinct cells, omit `max_atoms_per_system`, request distances/vectors plus `pair_fn`, compare selected full rows, and check a gradient.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,31 @@
 # Changelog
 
+## Unreleased
+
+### Added
+
+- JAX and Torch naive neighbor APIs now expose CUDA-only topology-only tiled compact
+  `target_indices` rows for no-PBC, wrapped PBC, and prewrapped PBC. For
+  native, Torch, and eager JAX single-system CUDA calls, partial topology-only
```
Comment (laserkelvin) [inline CHANGELOG.md:17]: issue: The unified auto selector still includes `has_target_indices` in `has_pair_outputs`, so `neighbor_list(method=None)` and `suggest_neighbor_list_method` remove `naive_tile` even for topology-only partial work. I assume the default API should receive this acceleration too. Split topology-only targets from geometry/pair outputs in `base_dispatch` while keeping the batching/rebuild guards.

File: nvalchemiops/torch/neighbors/naive.py [hunk 5]
```diff
@@ -1286,22 +1319,126 @@ def naive_neighbor_list(
...
+            num_neighbors,
+            neighbor_matrix_shifts,
+            fill_value=int(fill_value),
+            return_neighbor_list=return_neighbor_list,
+        )
+
+    uses_compact_pair_kernel = (
+        has_geometry_or_pair_outputs or target_indices is not None
+    )
+    if uses_compact_pair_kernel:
```
Comment (laserkelvin) [inline nvalchemiops/torch/neighbors/naive.py:1441]: issue: This compact geometry/pair branch runs before the later `cutoff <= 0` return, and the kernel squares the cutoff; `-1` therefore finds neighbors within radius 1 while topology returns empty. I assume all output modes share the nonpositive-cutoff contract. Reset and return before the topology/geometry split, then add single and batched Torch regressions.

File: nvalchemiops/torch/neighbors/batch_naive.py [hunk 6]
```diff
@@ -1477,22 +1637,10 @@ def batch_naive_neighbor_list(
...
+    uses_compact_pair_kernel = (
+        has_geometry_or_pair_outputs or target_indices is not None
     )
-    if strategy == "tile" and target_indices is not None:
-        raise NotImplementedError(
-            "strategy='tile' has no target_indices (partial "
-            "neighbor-list) variant; use strategy='scalar'.",
-        )
-    if has_pair_outputs:
+    if uses_compact_pair_kernel:
```
Comment (laserkelvin) [inline nvalchemiops/torch/neighbors/batch_naive.py:1643]: issue: Before this compact branch, missing companion buffers cause `max_neighbors` to be estimated even when `neighbor_matrix` supplies a custom width; validation then rejects that matrix, unlike the single-system API. I assume the supplied matrix width defines capacity. Infer `max_neighbors` from `neighbor_matrix.shape[1]` first, then size missing buffers from it.

File: test/neighbors/bindings/jax/test_batch_naive.py [hunk 7]
```diff
@@ -367,20 +392,154 @@ def test_target_indices_rejects_full_size_user_buffers(self):
...
+            ),
+            device,
+        )
+        batch_idx = jax.device_put(jnp.array([0, 0, 1, 1], dtype=jnp.int32), device)
+        batch_ptr = jax.device_put(jnp.array([0, 2, 4], dtype=jnp.int32), device)
+        target_indices = jax.device_put(jnp.array([2, 0], dtype=jnp.int32), device)
+        kwargs = {}
+        if pbc_mode != "none":
+            cell = jax.device_put(
+                jnp.stack([jnp.eye(3, dtype=dtype) * 4.0] * 2),
```
Comment (laserkelvin) [inline test/neighbors/bindings/jax/test_batch_naive.py:424]: suggestion: Both systems use identical cells/PBC and equal atom counts, so this parity test cannot catch selecting metadata by compact row instead of `target_indices[row]`. I assume per-target system lookup is part of this regression. Use different cells or PBC axes plus an out-of-cell coordinate, then compare independent `(neighbor, shift)` multisets.

File: nvalchemiops/jax/neighbors/naive.py [hunk 8]
```diff
@@ -2041,21 +2520,15 @@ def naive_neighbor_list(
...
-    silently degrades the wrapped path to cold-capture-per-call (correct,
-    but significantly slower than the proposal's measured replay numbers).
+    cache via ``warp.jax_experimental.set_jax_callable_default_graph_cache_max(...)``;
+    set that default before importing this module because its callables are
+    registered at import time.
+
+    ``graph_mode="warp"`` replay requires stable input and output buffers.
+    Donate the returned output buffers through the enclosing ``jax.jit`` so
+    they round-trip between calls. Wrapped PBC also requires stable
+    ``inv_cell``, ``positions_wrapped``, and ``per_atom_cell_offsets`` buffers.
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/naive.py:2175]: issue: These are deprecated aliases scheduled for removal in 0.5; the canonical names are `inv_cell_buffer`, `positions_wrapped_buffer`, and `per_atom_cell_offsets_buffer`. I assume new replay guidance should teach the stable API. Update this docstring, replay example, and related errors to the canonical names, and clarify compact buffer shapes/donation.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,31 @@
...
+  `auto` selects a strategy from the dtype and atom count. JAX traced partial
+  `auto` calls without `graph_mode="warp"` and batched partial `auto` calls
+  remain scalar; `graph_mode="warp"` routes partial `auto` to tile.
+  Explicit tile supports single and batched compact rows. Geometry and pair
+  outputs use scalar, and partial lists reject `rebuild_flags` under every
+  strategy. Scalar and tiled results have the same per-row neighbor and
```
Comment (laserkelvin) [inline CHANGELOG.md:23]: nit: This multiset-equivalence claim only holds when a row does not overflow; after `num_neighbors > max_neighbors`, scalar and tile may retain different capped subsets, not merely different ordering. I assume the changelog should distinguish uncapped counts from stored entries. Qualify the claim with “when no row overflows” and compare counts against capacity first.
